### PR TITLE
fix: improve dial queue and parallel dials

### PR DESCRIPTION
**Pull request by jacobheun**
## Improved Dialer
This improves the dialer queue a bit by doing a few things:

**Multidimensional queues**
As dial requests are made they no longer go into a single, global queue, they are immediately added to the dial queue for that specific peer. This allows a successful queue, or one that errors, to immediately resolve the remaining dialing requests. Previously, a subsequent request would need to wait until its turn globally, which is slow, especially if it just needs to be aborted or a new stream needs to be created for a protocol.

**Immediate dials for connected peers**
If a peer is already connected, we immediately start its queue if it's not already running. The max parallel queues aren't affected by this. This ensures that we have a steady stream of communication with our connected peers, which is critical to maintaining a healthy network.

## Configurable dial timeout
The dial timeout was also not configurable. This has been corrected and added to the readme.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,7 +62,8 @@ const sw = new switch(peerInfo , peerBook [, options])
 If defined, `options` should be an object with the following keys and respective values:
 
 - `blacklistTTL`: - number of ms a peer should not be dialable to after it errors. Defaults to `120000`(120 seconds)
-- `maxParallelDials` - number of concurrent dials the switch should allow. Defaults to `50`
+- `maxParallelDials`: - number of concurrent dials the switch should allow. Defaults to `50`
+- `dialTimeout`: - number of ms a dial to a peer should be allowed to run. Defaultd to `30000` (30 seconds)
```
Comment (alanshaw) [inline README.md]:
```suggestion
- `dialTimeout`: - number of ms a dial to a peer should be allowed to run. Defaults to `30000` (30 seconds)
```

File: src/limit-dialer/index.js [hunk 2]
```diff
@@ -41,20 +41,27 @@ class LimitDialer {
...
+          successfulDial = res
+        }
+
+        cb()
       })
     })
 
-    tryEach(tasks, (_, result) => {
-      if (result && result.conn) {
+    parallel(tasks, (_) => {
```
Comment (alanshaw) [inline src/limit-dialer/index.js]: Use `async/each` instead so you can share the task function?

```js
each(addrs, (m, cb) => { this.dialSingle(/*...*/) })
```

Comment (alanshaw) [inline src/limit-dialer/index.js]: Actually, do you want to use `async/race` for this instead so you don't have to wait for all the dials to finish before calling back?

File: src/limit-dialer/queue.js [hunk 3]
```diff
@@ -54,9 +54,9 @@ class DialQueue {
         pull(empty(), conn)
         // If we can close the connection, do it
         if (typeof conn.close === 'function') {
-          return conn.close((_) => callback(null, { cancel: true }))
+          return conn.close((_) => callback(null))
         }
-        return callback(null, { cancel: true })
+        return callback(null)
```
Comment (alanshaw) [inline src/limit-dialer/queue.js:59]: What does removing these objects do?

**Review by alanshaw — CHANGES_REQUESTED** [review]

File: src/limit-dialer/queue.js (hunk 3 above)
Comment (jacobheun) [reply src/limit-dialer/queue.js:59]: I believe these were left over from some old code, only valid connection results are being used now, so they were deleted. The intent is that it creates a cancel token that causes other successfully dialed connections to be destroyed. Now the token is created when we receive the first valid connection.

File: src/limit-dialer/index.js (hunk 2 above)
Comment (jacobheun) [reply src/limit-dialer/index.js]: I will use each. `race` will finish as soon as any of the dial callbacks finish, even if nothing is passed. So if we dial 5 addresses and the first errors, even if we just store an array of the errors as soon as we callback it ends, so we end up having a "failed" dial after the first one errors. Even if the other 4 were actually good.

Comment (jacobheun) [reply src/limit-dialer/index.js]: Okay, I did go with race for the aforementioned benefits. I just only call the callback if we have a value, or if everything errored.

**Review by alanshaw — APPROVED** [review]

**Pull request merged**